intro-to-oop-contd/testing-objects.md

```diff
@@ -0,0 +1,263 @@
+
+### Exercise: Test Cases
+
+Read through this project description. Then, make a list of all relevant test cases. This list should mirror Scarlet's list of tests for her ride share app.
+
+Imagine an app that manages the customers and orders of a grocery store. There are two classes, `Customer` and `Order`.
+
+Each `Customer` has:
+    - `name`
+    - `email_address`
+
+Each `Order` has:
+    - `products`
+    - `customer`
+    - `add_product` method
+    - `remove_product` method
+
+Imagine and make up assumptions about the logic of `add_product` and `remove_product`, particularly around the **edge cases**.
+
+<!-- Question 1 -->
+<!-- prettier-ignore-start -->
+### !challenge
+
+* type: short-answer
+* id: 993d92bb-2c3d-43cd-b6cc-7268c0b63b71
+* title: test cases
+
+##### !question
+
+From the given project description, make a list of all relevant test cases. Imagine and make up assumptions about the logic of `add_product` and `remove_product`, particularly around the **edge cases**.
+
+##### !end-question
+
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+## Revisit pytest Syntax
+
+Our pytest unit tests will test a class's state and behavior. However, overall, our syntax does not change very much.
+
+```python
+def test_some_example_test_case():
+    # Arrange
+    # Create an instance of the class
+    # and set up any other necessary test variables
+
+    # Act
+    # Call the method that we are testing
+
+    # Assert
+    # Verify all relevant return values and state changes
+```
+
+When testing the initial state of an instance, the "Act" step is likely the instantiation step, instead of "Arrange".
+
+### Example
+
+Let's dive immediately back into Scarlet's ride share app, and her tests for `Driver`. In Scarlet's app, imagine there is a `Trip` class that has an optional `rating` attribute.
+
+This test tests a driver's initial state:
+
+```python
+def test_new_valid_driver():
+    name = "Batman"
+    vin = "NAN4NAN4NA"
+    trip_a = Trip()
+    trip_b = Trip()
+    trips = [trip_a, trip_b]
+
+    batman = Driver(name=name, vin=vin, trips=trips)
+
+    assert batman.name is name
+    assert batman.vin is vin
+    assert len(batman.trips) is 2
+    assert trip_a in batman.trips
+    assert trip_b in batman.trips
+```
+
+We should notice:
+
+- In our Arrange step, we'll set up the `name`, `vin`, `trip_a`, `trip_b`, and `trips` variables
+- In our Act step, we instantiate `batman`
+- In our Assert step, we check:
+    - `batman`'s `name` attr
+    - `batman`'s `vin` attr
+    - The length of `batman`'s `trips`
+    - The contents of `batman`'s `trips`
+
+### !callout-info
+
+## Why is Checking the `trips` List So Weird?
+
+In the above test, we checked that `batman.name` is equal to `"Batman"`. However, for `batman.trips`, we checked the length and the contents of the trip. Python does not check equality between lists like how a human might imagine. Our best strategy is to check the length and contents.
```

Comment:
I think I probably mistakenly assumed something; I'll read back through and see if I can cut this callout and adjust the tests if needed

Reply:
I agree with you that the content here is wrong.

Now here's a code style thing, if you'd like to throw your wisdom in here:

```python
    assert len(batman.trips) == 2
    assert trip_a in batman.trips
    assert trip_b in batman.trips
```

vs.

```python
    assert batman.trips == [trip_a, trip_b]
```


I'd argue that approach #1 is more flexible (depending on how `in` works, doesn't necessitate `trips` being a list), but I don't know enough Python to know how much that matters.

Reply:
`in` calls `__contains__`, so it requires that the object be something that implements `__contains__`.  In general code, I could understand wanting to allow for different data types (though usually swapping collection types is something we do at more than just the drop of a hat), so having the test encode that the expected collection type doesn't seem overly strict.  There's also the question of whether the collection is ordered or not and whether that's important.  But all of those concerns seem a bit esoteric to me in this context.

If I were writing this test, I probably wouldn't even bother creating a new array for the assertion.  I think
```python
    assert batman.trips == trips
```
would be fine (not _is_ trips, since the ctor might duplicate the list).  But if the thing coming back is indexable and so forth, might as well just own the fact that it's probably a list.  Of course, even just exposing arrays of values when they are so easily modified without the owner knowing is a whole discussion on its own!